File: array/MatrixMultiplication/cpp/MatrixMultiplication.cpp
#include "MatrixMultiplication.hpp"

#include <stdexcept>

using namespace std;

Matrix multiplyMatrices(const Matrix& left, const Matrix& right) {
    if (left.empty() || right.empty()) {
        throw invalid_argument("matrices must be non-empty");
    }

    const size_t rows = left.size();
    const size_t shared = left.front().size();
    const size_t rightRows = right.size();
    const size_t cols = right.front().size();

    if (shared == 0 || cols == 0 || shared != rightRows) {
        throw invalid_argument("matrix dimensions are incompatible");
    }

    for (const auto& row : left) {
        if (row.size() != shared) {
            throw invalid_argument("left matrix rows have inconsistent lengths");
        }
    }
    for (const auto& row : right) {
        if (row.size() != cols) {
            throw invalid_argument("right matrix rows have inconsistent lengths");
        }
    }

    Matrix result(rows, vector<long long>(cols, 0));
    for (size_t i = 0; i < rows; ++i) {
        for (size_t k = 0; k < shared; ++k) {
            const long long value = left[i][k];
            for (size_t j = 0; j < cols; ++j) {
                result[i][j] += value * right[k][j];
            }
        }
    }
    return result;
}

File: array/MatrixMultiplication/java/MatrixMultiplication.java
public final class MatrixMultiplication {
    private MatrixMultiplication() {}

    public static long[][] multiply(long[][] left, long[][] right) {
        if (left.length == 0 || right.length == 0) {
            throw new IllegalArgumentException("matrices must be non-empty");
        }

        int rows = left.length;
        int shared = left[0].length;
        int rightRows = right.length;
        int cols = right[0].length;

        if (shared == 0 || cols == 0 || shared != rightRows) {
            throw new IllegalArgumentException("matrix dimensions are incompatible");
        }

        for (long[] row : left) {
            if (row.length != shared) {
                throw new IllegalArgumentException("left matrix rows have inconsistent lengths");
            }
        }
        for (long[] row : right) {
            if (row.length != cols) {
                throw new IllegalArgumentException("right matrix rows have inconsistent lengths");
            }
        }

        long[][] result = new long[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int k = 0; k < shared; k++) {
                long value = left[i][k];
                for (int j = 0; j < cols; j++) {
                    result[i][j] += value * right[k][j];
                }
            }
        }
        return result;
    }
}
